feat: Pinecone document store optimizations

ivanazeljkovic opened the pull request:
### Related Issues

Resolves: [Pinecone Document Store optimizations #5901](https://github.com/deepset-ai/haystack/issues/5901)

### Proposed Changes

Optimize `delete_documents()` and `_get_vector_count()` methods from `PineconeDocumentStore`.

### How did you test it?

Testing is done through existing unit and integration tests, as well as manually.

### Checklist

- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [x] I have updated the related issue with new insights and changes
- [x] I added unit tests and updated the docstrings
- [x] I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: fix:, feat:, build:, chore:, ci:, docs:, style:, refactor:, perf:, test:.
- [x] I documented my code
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

anakin87 commented:
Please merge main/rebase on main...
We fixed a test error in #5906.

silvanocerza commented on haystack/document_stores/pinecone.py:
```diff
@@ -865,39 +905,50 @@ def _get_all_document_ids(
+                            batch_size=batch_size,
+                        )
+                        # Save IDs
+                        all_ids = all_ids.union(set(vector_id_matrix))  # type: ignore
+                        # Move these IDs to new namespace
+                        self._move_documents_by_id_namespace(
+                            ids=vector_id_matrix,
+                            index=index,
+                            source_namespace=namespace,
+                            target_namespace=target_namespace,
+                            batch_size=batch_size,
+                        )
+                        progress_bar.set_description_str("Retrieved IDs")
+                        progress_bar.update(len(set(vector_id_matrix)))
```
We can avoid recreating the `set` multiple times this way.
```suggestion
                        unique_ids = set(vector_id_matrix)
                        all_ids = all_ids.union(unique_ids)  # type: ignore
                        # Move these IDs to new namespace
                        self._move_documents_by_id_namespace(
                            ids=vector_id_matrix,
                            index=index,
                            source_namespace=namespace,
                            target_namespace=target_namespace,
                            batch_size=batch_size,
                        )
                        progress_bar.set_description_str("Retrieved IDs")
                        progress_bar.update(len(unique_ids))
```

silvanocerza reviewed (changes requested):
Looks great, just a couple of changes to make and it's ready to go. 👍

julian-risch reviewed (dismissed):
Thanks for opening this pull request! I just have two small change requests in addition to the change request by Silvano. Otherwise the pull request looks good to me! 👍

ivanazeljkovic commented:
@silvanocerza @julian-risch Thank you for a detailed review and feedback! 
I fixed the things you pointed out. Also, improved few more things in the meantime, so your feedback is very welcome. 

Improvements:
- _get_embedding_count_ method - if filters include `doc_type` related one, it will be removed and replaced with appropriate one (`doc_type: "vector"`), since this one is the only valid filter related to this meta field when counting documents with embedding,
- _get_document_count_ method - default value for `doc_type` meta field will be added only if filters doesn't include `doc_type` related ones,
- additional optimizations related to deleting documents (refactoring underlying methods and increasing default batch size)
- labels retrieval fix

silvanocerza commented:
Hey @ivanazeljkovic, I see there are some issues with CI. 
They should have been fixed in `main` but I can't edit your PR, can you please update it so that CI runs again?

ivanazeljkovic commented:
Hi @silvanocerza, I have done a rebase so now all tests are passing successfully.

silvanocerza reviewed (approved):
Nice! 🚀
